### ISSUE TITLE
Having issues running your code

### ISSUE THREAD

**Issue by igordertigor** [eldrin/MTLMusicRepresentation-PyTorch#1]
Hi Eldrin,

after reading your paper, I was hoping that I could play with your code a bit, but that appears quite difficult. I'm not entirely sure if a github issue is the best way for this, please let me know if you prefer a different mode of contact. I tried to start by running the `scripts/ext_feat.py` script that you mention in the readme. Here is a list of the different issues I'm facing:

1. torch==1.13.0+cu116 has no installation candidate, so I modified the requirements.txt to only refer to torch==1.13.0.
2. Next, the definition of `DEFAULT_SCALER_REF` in `musmtl/model.py` line 9 is broken. I'm getting an `AttributeError: module 'importlib.resources' has no attribute 'files'`. However, it appears that pretty much no other place in the code is referencing this, except for line 327 in `musmtl/tool.py` and that use is inside a main function that isn't used by the `scripts/ext_feat.py` script. So I decided to set `DEFAULT_SCALER_REF=None` and move on.
3. That gives me a couple of `TypeError`s in `musmtl/utils.py`. All of them are related to type hints and are easy to fix (`from typing import Tuple, List` and then `tuple[...] -> Tuple[...], list[...] -> List[...]`).
4. However, then I run into an issue that I am not able to resolve: `musmtl/tool.py` on line 92 uses `np.load` on the `scaler_fn`, which points to the file `models/sclr_dbmel.dat.gz`. Here, I get the following error message:
```
Traceback (most recent call last):
  File "scripts/ext_feat.py", line 47, in <module>
    for feature, fn in zip(ext.run(model_fns, SCALER_FN), model_fns):
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/scripts/../musmtl/tool.py", line 163, in run
    scaler, model = self._load_model(model_fn, scaler_fn, self.device)
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/scripts/../musmtl/tool.py", line 92, in _load_model
    with np.load(scaler_fn) as npf:
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/.venv/lib/python3.8/site-packages/numpy/lib/npyio.py", line 445, in load
    raise ValueError("Cannot load file containing pickled data "
ValueError: Cannot load file containing pickled data when allow_pickle=False
```
Of course, I tried to set `allow_pickle=True`, but the file doesn't appear to actually be a pickle file:
```
Traceback (most recent call last):
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/.venv/lib/python3.8/site-packages/numpy/lib/npyio.py", line 448, in load
    return pickle.load(fid, **pickle_kwargs)
_pickle.UnpicklingError: invalid load key, '\x1f'.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "scripts/ext_feat.py", line 47, in <module>
    for feature, fn in zip(ext.run(model_fns, SCALER_FN), model_fns):
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/scripts/../musmtl/tool.py", line 163, in run
    scaler, model = self._load_model(model_fn, scaler_fn, self.device)
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/scripts/../musmtl/tool.py", line 92, in _load_model
    with np.load(scaler_fn, allow_pickle=True) as npf:
  File "/home/ingo/code/MTLMusicRepresentation-PyTorch/.venv/lib/python3.8/site-packages/numpy/lib/npyio.py", line 450, in load
    raise IOError(
OSError: Failed to interpret file 'scripts/../models/sclr_dbmel.dat.gz' as a pickle
```
I have a feeling that it might be related to line 91 in `musmtl/tool.py`, where a joblib load command is commented out, but I don't know how to fit the two together. Can you give me a pointer?

Would it help if I made a pull request with my changes?

Also, I find it very difficult to make sense of the different checkpoints in the data.zip file that the readme references. There are almost 400 checkpoint folders in `mtl_dataset/mtlmodel_release`. Which one is good? Which one(s) is the paper based on?

**Comment by eldrin** [comment]
Hi igordertigor. The issue board is absolutely fine to discuss this. thank you for bringing it up!

I am unfortunately not able to test the whole pipeline right now (will need a week or so to be able to check myself due the circumstance). I have a few ideas though that you might need to try in between.

I believe the 4 issues are brought from the version mismatch.

Did you try to install the virtual env using the `environment.yaml` as documented in `README.md`? I at least tested this and my Linux machine (Ubuntu 20.04) doesn't have a problem installing the environment. And I believe most of the issues you listed from 1 ~ 3 would be resolved by that. (seems you've installed it on python older than python3.9 and maybe using pip)

As for issue 4, please try `models/sclr_dbmel3.dat.gz` after installing the environment above way. I believe that this one is updated due to that issue.

As for the last part, I would say the conclusion of the study was that "**there is no absolute winner, but some may be conditionally better than others.**" thus the question always should be "**good for what?**". When you consider some model is good, what do you expect the model to do good for?

the about 400 models are the models that we populated to answer the question, which still is sampled from more than a thousand models that we could populate to get a more robust answer, which unfortunately we couldn't due to the limited access to the GPUs back in the day.

We do have a few general trends correlated to something that might be close to the "general goodness", so that comprises our conclusion / general recommendations:

1. more tasks would be more useful
2. architecture that shares fewer layers would be more useful
3. as for the single task winner, the `artist-tag` one generally works well with various tasks, but not always.

Considering the above criteria, you would be able to find the model checkpoint that is within them. I can't recall the one from my memory right now so I will also give you an update when I can test the extraction part myself.

I hope that this at least partially resolved your issues. I will get back to you as soon as possible.

**Comment by igordertigor** [comment]
Awesome, thank you @eldrin, yes, I installed everything using pip with py3.8 (kind of the default on my system). Anything else would mean upgrading python and installing conda. I'll need to spend a bit of time getting these up and running (it also isn't really a priority at the moment) and then I'll try the `artist-tag` model. However, I think my "issue 4" was rather: How do I map the names from the paper to the names in the models file?

**Comment by eldrin** [comment]
No problem! Let me know if you have any further issue with the installation later.

As for your question: we designed our experiment by the on-demand version of [fractional factorial design](https://en.wikipedia.org/wiki/Fractional_factorial_design). (For the details, check 4.3 of the paper where we described our experimental design). Computed experimental design can be found in [this file](https://github.com/eldrin/MTLMusicRepresentation-PyTorch/blob/5dc6d66df8d028faa42e2c27fb4c3e4946501887/eval/data/x.csv).

Each row of the csv file represents a trial with various multiple learning sources and the "branching" point of the architecture. The first 8 columns represents each learning source (self, bpm, year, taste, tag, lyrics, cdr_tag, artist) and the 9th column represents the code for the architecture setup.

 https://github.com/eldrin/MTLMusicRepresentation-PyTorch/blob/5dc6d66df8d028faa42e2c27fb4c3e4946501887/scripts/post_process_results.py#L16-L25

Each model checkpoint maps to each of the experimental trial based on this design matrix. (eg. 0_199.pth is the first row of the design, 399_199.th is the 400th row of the design)
